add (e)grep / no_(e)grep matcher

schorsch3000 opened the pull request:
# What?

This pullrequest addresses both parts of #96 
it fixes the documentation in the `README.md` and brings regex_matcher using grep and egrep
# Why?

I need regex matching to test output of shellscripts, using an extra variable und call grep or egrep in tests is inconvinient
# How was it tested?

with shpec of course, I've added test-cases for all 4 matchers.

schorsch3000 commented:
Hmm, it fails on zsh.
im not familliar with zsh.
any hints?

twe4ked commented:
Could be that Zsh uses brackets in it's pattern matching. The `[`, `]` may need to be escaped?

schorsch3000 commented:
i don't think that `  assert no_egrep "hello1" "^[a-z]+$"` needs to habe it's [] escaped where `       assert test "[ 5 -lt 10 ]"` is fine, do you?

schorsch3000 commented:
Sure, that parameters should be quoted, they get evaled...
that'll fix it :)

rylnd commented:
@schorsch3000 resurrecting [my previous comment](https://github.com/rylnd/shpec/issues/96#issuecomment-277566694): could you explain the value of having both `match` and `grep` matchers? I'm not sure that we need both, but if we do we need to better document how to use them: the existing `match` tests would also pass using both the `grep` and `egrep` matchers.

rylnd commented on bin/shpec:
```diff
@@ -107,6 +107,18 @@ assert() {
       "Expected $2 to be true"
   ;;
+  ( xgrep )
+    print_result "echo \"$2\" | grep -q \"$3\"" "Expected $2 to match $3"
```
For consistency with our other output (and, more importantly, for easier parsing by the user), we should wrap the arguments in `[]` here.

schorsch3000 commented:
So, i've aligned the messages to match the style, 
renamed match and no_match to glob and no_glob,
reworked the readme
and added some tests to show the difference between glob and grep.
that should fix #97 and #103 atm i've no idea to fix #104, but i will think about that since it bugs me too.

rylnd commented on bin/shpec:
```diff
@@ -74,11 +74,11 @@ assert() {
   ;;
-  ( xmatch )
+  ( xglob )
     print_result "case '$2' in *$3*) :;; *) false;; esac" \
```
@schorsch3000 I would prefer if we did away with the default globbing here; there should be no (potentially unexpected) behavior beyond what is passed to the matcher.

That is, `assert glob "partially" "partial"` should fail; if users want that behavior, they should use `assert grep "partially" "partial"` instead.

rylnd commented:
@schorsch3000 👍 👍 👍 , fix up the default globbing and this is good to merge.

rylnd commented:
@schorsch3000 save for the globbing issue (and, now, a rebase 😢 ), this is good to go. if you don't have the time/have lost interest/don't agree, please let me know. With your permission, I would be happy to do the remaining work on your behalf.

schorsch3000 commented:
Hi @rylnd 
lets face it,
i have the time to work on the globbing issue, in fact, i've fiddled around quite a bit.
I'm interested in this PR and in a fixed globbing.
I'm not sure if the globbing-issuse is in the scope of this PR or not, but i've tried a fix, no need to discuss that.

the problem is just that i have no idea how to solve the issue, that seems to be just above my head :(
If you like to work on that and know a way to fix that, i would be happy.

If i can help somehow and you point me in the right direction, i'm happy to do that as well.

Thanks

rylnd commented:
@schorsch3000 sorry for the confusion! It was easier to make the changes than it was to try to convey my intent to you. Those are on a new branch: [glob_without_wildcards](https://github.com/rylnd/shpec/tree/glob_without_wildcards).

Feel free to cherry-pick my commits to your branch, or use that branch to open a new PR, or whatever you want to do to get this work resolved.

schorsch3000 commented:
Oh, nice!
...and it merged without conflicts.
Are we good to go?

rylnd commented:
@schorsch3000 yes! Thank you for your help on this.

schorsch3000 commented:
I'm glad its done, thanks! :)
